### ISSUE TITLE
Architecture?

### ISSUE THREAD

**Issue by hmijail** [ConsenSys/evm-dafny#508]
The ARCHITECTURE.md doc needs a revision, and I'd suggest that we should find something more informative than "hi", "mid" and "low" layers. (unless that's some kind of standard that I'm missing?)

It might be useful to put things in terms of the discussion of "EVM equivalence".  I suspect that the "high layer" is actually the Execution layer, and the "mid layer" is the actual EVM.
https://ethereum-magicians.org/t/evm-equivalence-eip-6269-and-ethereum-stack-compatibility-definition/10044/9

**Comment by hmijail** [comment]
Also, ARCHITECTURE.md (why is it in caps?) includes a graph but AFAICT the repo doesn't include whatever is necessary to rebuild it.

**Comment by DavePearce** [comment]
Well, I agree the `ARCHITECTURE.md` can be improved.  There's a lot more we could say here, including some stuff already written in the long version of the FM paper.

Also agree that we could consider other terminology instead of "high", "mid", "low", etc.  But, this is really an internal description that is not connected (as far as I'm aware) with anything external.  Everything is really just the EVM itself.  I don't think we have much (if anything) specific for the execution layer.  Perhaps some things on the Java side?  However, if there is some useful terminology, that would be nice.

**Comment by DavePearce** [comment]
> Also, ARCHITECTURE.md (why is it in caps?)

Well, just following conventions.  Like, e.g. `README.md` and `CONTRIBUTORS.md`, etc.

**Comment by DavePearce** [comment]
> includes a graph but AFAICT the repo doesn't include whatever is necessary to rebuild it.

So, its an an xfig file (`.fig` extension).  We could include that I guess.  Maybe there is some preferred format though?  Xfig is ... old.  And, I tend to use it when writing papers ... but only because I've been doing that for the last 20 years :)

**Comment by hmijail** [comment]
The ALLCAPS naming is a [Unix vestige/tradition in basic root-of-repo files like README, INSTALL, CONTRIBUTORS, etc](https://softwareengineering.stackexchange.com/questions/301691/readme-txt-vs-readme-txt). Not sure it makes sense to use it for other random files (which probably should go into a docs/ directory or similar?)

**Comment by DavePearce** [comment]
Exactly.  I wouldn't consider it a random file though.  See e.g. [here](https://changelog.com/news/why-you-should-have-an-architecturemd-ARmA) and [here](https://matklad.github.io/2021/02/06/ARCHITECTURE.md.html).

**Comment by hmijail** [comment]
Link 2 is just a recap of link 1, so both give the same single example. 🤔
And then we have SEMANTICS, etc...

**Comment by DavePearce** [comment]
I'm sure I can find more examples easily enough.  I think its fine as is.  As long as we're consistent, then no problem.  Anything ending in `.md` is in caps ... easy.

**Comment by DavePearce** [comment]
Randomly, I've just been reading [the KEVM paper](https://core.ac.uk/download/pdf/158321354.pdf) (which I should have done before now) and noticed it splits into three levels:

![KFrameworkLevels](https://user-images.githubusercontent.com/484459/211704388-73b0f5d2-30eb-48e8-879e-287d01512f11.png)

I'd say our high level is similar to their `evm.k` level, and perhaps our mid and low levels compare to their `data.k` level.  In some sense, our low level is just a bunch of utils really.
